File: backend/app/schemas/ecu_merger.py
from pydantic import BaseModel, Field

from app.schemas.generation import (
    GeneratedRequirementCandidateBase,
    GeneratedRequirementCandidateValidation,
)
from app.schemas.requirement import RequirementResponse


class EcuRequirementMergerAnalyzeRequest(BaseModel):
    project_id: str = Field(min_length=1, max_length=64)
    requirement_ids: list[str] = Field(min_length=1)


class EcuRequirementMergerTraceabilityItem(BaseModel):
    requirement: RequirementResponse
    source_subsystem: str | None = None
    source_feature_label: str | None = None
    traceability_reason: str


class EcuRequirementMergerAffectedDesignParameter(BaseModel):
    id: str
    name: str
    parameter_name: str
    value: str
    unit: str | None = None
    subsystem: str | None = None
    linked_requirement_ids: list[str] = Field(default_factory=list)
    reason: str


class EcuRequirementMergerCandidateReview(GeneratedRequirementCandidateBase):
    temp_id: str = Field(min_length=1, max_length=64)
    suggested_id: str = Field(min_length=1, max_length=64)
    suggested_hierarchy: str | None = Field(default=None, max_length=128)
    validation: GeneratedRequirementCandidateValidation
    draft_label: str
    source_requirement_ids: list[str] = Field(default_factory=list)
    source_requirement_codes: list[str] = Field(default_factory=list)
    source_subsystems: list[str] = Field(default_factory=list)
    merge_rationale: str
    assumptions_list: list[str] = Field(default_factory=list)
    conflicts: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    recommended_fix_actions: list[str] = Field(default_factory=list)
    verification_notes: list[str] = Field(default_factory=list)
    shared_signals_or_interfaces: list[str] = Field(default_factory=list)
    timing_constraints: list[str] = Field(default_factory=list)
    safety_security_constraints: list[str] = Field(default_factory=list)
    affected_design_parameters: list[EcuRequirementMergerAffectedDesignParameter] = Field(default_factory=list)
    traceability: list[EcuRequirementMergerTraceabilityItem] = Field(default_factory=list)


class EcuRequirementMergerAnalyzeResponse(BaseModel):
    project_id: str
    selected_requirement_ids: list[str] = Field(default_factory=list)
    selected_requirement_count: int
    candidates: list[EcuRequirementMergerCandidateReview] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class EcuRequirementMergerSaveCandidate(GeneratedRequirementCandidateBase):
    temp_id: str = Field(min_length=1, max_length=64)


class EcuRequirementMergerSaveRequest(BaseModel):
    project_id: str = Field(min_length=1, max_length=64)
    candidates: list[EcuRequirementMergerSaveCandidate] = Field(min_length=1)

File: frontend/src/types/ecuMerger.ts
import type {
  Requirement,
  RequirementCorrelationSummary,
  RequirementGenerationMetadata,
  RequirementPriority,
  RequirementQualitySummary,
  RequirementType,
  StructuredRequirementFields,
} from "./requirements";

export type EcuRequirementMergerCandidateValidation = {
  quality_summary: RequirementQualitySummary;
  parsed_requirement: StructuredRequirementFields;
  correlation_summary: RequirementCorrelationSummary;
};

export type EcuRequirementMergerTraceabilityItem = {
  requirement: Requirement;
  source_subsystem: string | null;
  source_feature_label: string | null;
  traceability_reason: string;
};

export type EcuRequirementMergerAffectedDesignParameter = {
  id: string;
  name: string;
  parameter_name: string;
  value: string;
  unit: string | null;
  subsystem: string | null;
  linked_requirement_ids: string[];
  reason: string;
};

export type EcuRequirementMergerCandidateReview = {
  temp_id: string;
  suggested_id: string;
  suggested_hierarchy: string | null;
  title: string;
  text: string;
  type: RequirementType;
  priority: RequirementPriority;
  rationale: string | null;
  parent_requirement_id: string | null;
  subsystem: string | null;
  verification_method: string | null;
  assumptions: string | null;
  generation_metadata: RequirementGenerationMetadata;
  validation: EcuRequirementMergerCandidateValidation;
  draft_label: string;
  source_requirement_ids: string[];
  source_requirement_codes: string[];
  source_subsystems: string[];
  merge_rationale: string;
  assumptions_list: string[];
  conflicts: string[];
  warnings: string[];
  recommended_fix_actions: string[];
  verification_notes: string[];
  shared_signals_or_interfaces: string[];
  timing_constraints: string[];
  safety_security_constraints: string[];
  affected_design_parameters: EcuRequirementMergerAffectedDesignParameter[];
  traceability: EcuRequirementMergerTraceabilityItem[];
};

export type EcuRequirementMergerAnalyzeRequest = {
  project_id: string;
  requirement_ids: string[];
};

export type EcuRequirementMergerAnalyzeResponse = {
  project_id: string;
  selected_requirement_ids: string[];
  selected_requirement_count: number;
  candidates: EcuRequirementMergerCandidateReview[];
  warnings: string[];
};

export type EcuRequirementMergerSaveCandidate = {
  temp_id: string;
  title: string;
  text: string;
  type: RequirementType;
  priority: RequirementPriority;
  rationale: string | null;
  parent_requirement_id: string | null;
  subsystem: string | null;
  verification_method: string | null;
  assumptions: string | null;
  generation_metadata: RequirementGenerationMetadata;
};

export type EcuRequirementMergerSaveRequest = {
  project_id: string;
  candidates: EcuRequirementMergerSaveCandidate[];
};

export type EcuRequirementMergerDraftState = EcuRequirementMergerCandidateReview & {
  draft_id: string;
  decision: "accepted" | "rejected";
};
